### ISSUE TITLE
'Tensor' object has no attribute 'head_transform1'

### ISSUE THREAD

**Issue by Shawnmhy** [deepfindr/gnn-project#1]
Hello,

I tried to replicate the codes in your youtube video 'GNN Project #3.1' and I got this error. Could you please suggest how to fix it? Thank you

**Comment by deepfindr** [comment]
Hi!
This certainly sounds like it's related to the transformerConv layer.

Which version of pytorch geometric are you using?

Also could you please paste the full error trace :) thx

**Comment by Shawnmhy** [comment]
> Hi! This certainly sounds like it's related to the transformerConv layer.
> 
> Which version of pytorch geometric are you using?
> 
> Also could you please paste the full error trace :) thx

Hi! I solved the problem, turns out I messed with a parameter name. I got the codes run properly, but the results seems very weird, the predictions are all positive labels. The confusion matrices are like:

[123 45]
[0   0   ]
for every epoch.

What do you think might be the cause?

Thank you for the reply.

**Comment by deepfindr** [comment]
OK great :)
The number of data points is a bit small. I remember that I had thousands of samples both for train and test. Are you sure that the data preprocessing worked correctly?

**Comment by Shawnmhy** [comment]
> 

Yes indeed the data points is a bit small, but I still don't believe it can predict 1 label and generate identical confusion matrices across all epochs. I wondering if I should change some parameters. I attached the question link with the code here:. 
[https://stackoverflow.com/questions/70028389/pytorch-geometric-gnn-model-only-predict-one-label](url)
Could you please take a look? Much appreciated!

**Comment by deepfindr** [comment]
Hi,

How many samples of the minority class do you have in your dataset?

**Comment by deepfindr** [comment]
Hi,

Did you oversample your data or use a weighted loss?

I had exactly the same problem at first. I also talk about this in the videos I uploaded.

In your case I think it's just not enough data and the best the model can do is to predict the class of which it has seen more, so always 0.

Best regards

**Comment by deepfindr** [comment]
By the way - there are already plenty of posts around that topic:
https://stackoverflow.com/questions/41488279/neural-network-always-predicts-the-same-class

In my experience, this often happens for inbalanced datasets. You should put more weight on your minority class or oversample it :)

**Comment by Shawnmhy** [comment]
> By the way - there are already plenty of posts around that topic: https://stackoverflow.com/questions/41488279/neural-network-always-predicts-the-same-class
> 
> In my experience, this often happens for inbalanced datasets. You should put more weight on your minority class or oversample it :)

Thank you for the suggestion! In fact, my dataset, through small, is quite balanced. There are 142 of label 1 and 151 of label 0. But I also put more weight on the minority class just in case. I tried to decrease the learning rate to very small: 0.000001. Now the model is predicting 'normally' in the training set, that is I can see some 1 being predicted. However, in the testing set it is still imbalanced, the model is still always predicting 1. Any thoughts?

**Comment by deepfindr** [comment]
Hi,

That sounds like classical overfitting I think. So basically the model memorizes the training samples of class 1 and when it sees new data it's unable to distinguish them.

Do you use dropout? If so, maybe increase the dropout rate. You could also play around with the size of the network. Maybe it is too small or too big, so try to play around with the hidden dim size.

I had exactly the same problems for my dataset and that's why I set up a hyperparameter search in the code. Just put in some values you want to try out and it will do the job for you :) this lead me to the parameter combination with the best results.

:)
